### ISSUE TITLE
Return action values in contextual bandits

### ISSUE THREAD

**Issue by Kmeco** [VowpalWabbit/vowpal_wabbit#1808]
Hi everyone, 

we would like to use vw in our project and would really appreciate your advice! We'd like to use the --cb reduction, but need to get all the action values. The reason for this is that in some cases the top action isn't available so we'd like to use the second best option.

If I use the --audit flag, it prints all the action values but I can not seem to find a way how to get these returned separately without all the weights - parsing them is too expensive. We are currently working with the python wrapper. I am happy to help implement this myself and submit a pull request as I am sure others will find this useful in the future, but I am not sure where to start. 

I appreciate any suggestions!

**Comment by jackgerrits** [comment]
It looks like the --cb reduction is using the multiclass prediction type. @marco-rossi29 are you aware of a way the ranking can be known?

**Comment by marco-rossi29** [comment]
@Kmeco I think the simplest way to go is to use --cb_adf rather than --cb. Then you can see the ranking returned I guess the same way we have in Decision Service. In your case it can actually make sense to use the Decision Service directly rather than only VW

**Comment by Kmeco** [comment]
Thanks for the suggestion @marco-rossi29!
We already have the pipeline which takes care of tracking, rewards joining and retraining so we don't really need all the functions that Decision Service offers. All we would really need is to rank all the actions. 

I tried using --cb_adf but am having the same issue as described here #1346 when using a string input.

When I try to pass in a list of vw.example objects as such:
```
model = pyvw.vw(cb_explore_adf=True)
ex1 = [model.example("| a b d"), model.example("0:-5:0.5 | a b c")]
model.learn(ex1)
```
I get the following error
`TypeError: expecting string or example object as ec argument for learn, got <class 'list'>`

I also found a test under `vowpal_wabbit/python/tests/test_pyvw.py` which tests this type of input so I don't really understand what's going wrong. Any ideas? 
I am running the latest `vowpalwabbit==8.5.0` with python 3.6 in an ubuntu docker image. 

Anyways, even if I managed to get that running, there is still this issue #1784 which would prevent us from using the python wrapper... we need to have a prototype up and running within a week, so we're really looking for the easies / quickest solution :) 

I am either considering running the vw as a deamon and using the comand line interface with  `--cb_explore_adf` or implementing a method to pass on all action values to the python wrapper under `--cb_explore`. They are already being displayed when using --audit. Do you have an idea how to do this please? Honestly, I am pretty lost in the C++ code right now but would love to contribute!

**Comment by jackgerrits** [comment]
I see, the problem is that the 8.5.0 bindings do not support multiline examples which is what is needed here. I will see if we can get the bindings updated to at least 8.6.1 on PyPi.

In the meantime can you build and use the bindings from source? Are you using Windows or Linux?

**Comment by marco-rossi29** [comment]
@Kmeco Do you want to set up some meeting to understand your guys need and show you what we currently have in the Decision Service?

**Comment by Kmeco** [comment]
@jackgerrits I understand now, will install the bindings to the latest vw version :) there is still this issue however #1784, where the actions order isn't being returned with the results. What do you think about calling the command line interface from python as a workaround? described here: 
https://stackoverflow.com/questions/47045769/interacting-with-vw-daemon-using-python

@marco-rossi29 would love to set up a meeting and chat about what you've managed to achieve with the Decision Service! email me at jakub.kmec@exponea.com and we can agree on some time that suits you :)

**Comment by jackgerrits** [comment]
There is a fix incoming for #1784, so once that's in you should be unblocked on the Python side. I think you're better off consuming the bindings directly rather than using daemon mode

**Comment by jackgerrits** [comment]
@Kmeco I am going to close this, if this is still an issue please reopen the issue.
